linera-rpc/src/grpc_network.rs

```diff
@@ -351,60 +351,68 @@ where
-                continue;
-            }
+        receiver
+            .for_each_concurrent(None, |(cross_chain_request, shard_id)| {
+                let shard = network.shard(shard_id);
+                let remote_address = format!("http://{}", shard.address());
 
-            let shard = network.shard(shard_id);
-            let remote_address = format!("http://{}", shard.address());
-
-            // Send the cross-chain query and retry if needed.
-            for i in 0..cross_chain_max_retries {
-                // Delay increases linearly with the attempt number.
-                tokio::time::sleep(cross_chain_sender_delay + cross_chain_retry_delay * i).await;
-
-                let result = || async {
-                    let cross_chain_request = cross_chain_request.clone().try_into()?;
-                    let request = Request::new(cross_chain_request);
-                    let mut client =
-                        ValidatorWorkerClient::new(pool.channel(remote_address.clone())?);
-                    let response = client.handle_cross_chain_request(request).await?;
-                    Ok::<_, anyhow::Error>(response)
-                };
-                match result().await {
-                    Err(error) => {
-                        warn!(
-                            nickname,
-                            %error,
-                            i,
-                            from_shard = this_shard,
-                            to_shard = shard_id,
-                            "Failed to send cross-chain query",
-                        );
+                let pool = pool.clone();
+                let nickname = nickname.clone();
+
+                // Send the cross-chain query and retry if needed.
+                async move {
+                    if rand::thread_rng().gen::<f32>() < cross_chain_sender_failure_rate {
+                        warn!("Dropped 1 cross-chain message intentionally.");
+                        return;
                     }
-                    _ => {
-                        debug!(
+
+                    for i in 0..cross_chain_max_retries {
+                        // Delay increases linearly with the attempt number.
+                        tokio::time::sleep(cross_chain_sender_delay + cross_chain_retry_delay * i)
+                            .await;
+
+                        let result = || async {
+                            let cross_chain_request = cross_chain_request.clone().try_into()?;
+                            let request = Request::new(cross_chain_request);
+                            let mut client =
+                                ValidatorWorkerClient::new(pool.channel(remote_address.clone())?);
+                            let response = client.handle_cross_chain_request(request).await?;
+                            Ok::<_, anyhow::Error>(response)
+                        };
+                        match result().await {
```

Comment:
```suggestion
                        let cross_chain_request = cross_chain_request.clone().try_into()?;
                        let request = Request::new(cross_chain_request);
                        let mut client =
                            ValidatorWorkerClient::new(pool.channel(remote_address.clone())?);
                        let result = client
                            .handle_cross_chain_request(request)
                            .await
                            .map_err(anyhow::Error::from);
                        match result {
```

Reply:
The outer `async move` block can't return a result, so we can't use `?` here.

Reply:
Oh, now I see it.